raster/r.report/prt_json.c

```diff
@@ -0,0 +1,196 @@
+#include <stdio.h>
+#include "global.h"
+#include <grass/parson.h>
+#include <grass/glocale.h>
+
+JSON_Value *make_units(int ns, int nl)
+{
+    JSON_Value *units_value = json_value_init_array();
+    JSON_Array *units_array = json_array(units_value);
+    for (int i = 0; i < nunits; i++) {
+        int _ns = ns;
+
+        JSON_Value *unit_value = json_value_init_object();
+        JSON_Object *unit_object = json_object(unit_value);
+
+        if (unit[i].type == CELL_COUNTS) {
+            json_object_set_string(unit_object, "unit", "cell counts");
+            json_object_set_number(unit_object, "value", count_sum(&_ns, nl));
+        }
+        else if (unit[i].type == PERCENT_COVER) {
+            json_object_set_string(unit_object, "unit", "% cover");
+            int k = ns - 1;
+            while (k >= 0 && same_cats(k, ns, nl - 1))
+                k--;
+            k++;
+            double area = area_sum(&k, nl - 1);
+            area = 100.0 * area_sum(&_ns, nl) / area;
+            json_object_set_number(unit_object, "value", area);
+        }
+        else {
+            char *unit_name = NULL;
+            if (unit[i].type == ACRES) {
+                unit_name = "acres";
+            }
+            else if (unit[i].type == HECTARES) {
+                unit_name = "hectares";
+            }
+            else if (unit[i].type == SQ_MILES) {
+                unit_name = "square miles";
+            }
+            else if (unit[i].type == SQ_METERS) {
+                unit_name = "square meters";
+            }
+            else if (unit[i].type == SQ_KILOMETERS) {
+                unit_name = "square kilometers";
+            }
+            json_object_set_string(unit_object, "unit", unit_name);
+            json_object_set_number(unit_object, "value",
+                                   area_sum(&_ns, nl) * unit[i].factor);
+        }
+        json_array_append_value(units_array, unit_value);
+    }
+    return units_value;
+}
+
+JSON_Value *make_category(int ns, int nl, JSON_Value *sub_categories)
+{
+    JSON_Value *object_value = json_value_init_object();
+    JSON_Object *object = json_object(object_value);
+
+    CELL *cats = Gstats[ns].cats;
+    char *cp = construct_cat_label(nl, cats[nl]);
```

Comment:
How should ranged categories be represented?

Reply:
We could represent continuous datasets similarly to how it is already represented in the shell output to make things easier, and just add an additional range object.

For example here is the basic output for the `elevation` dataset.

```bash
r.report map=elevation@PERMANENT nsteps=3

+-----------------------------------------------------------------------------+
|                         RASTER MAP CATEGORY REPORT                          |
|PROJECT: nc_spm_08_grass7                            Mon Jul  1 11:51:57 2024|
|-----------------------------------------------------------------------------|
|          north: 228500    east: 645000                                      |
|REGION    south: 215000    west: 630000                                      |
|          res:       10    res:      10                                      |
|-----------------------------------------------------------------------------|
|MASK: none                                                                   |
|-----------------------------------------------------------------------------|
|MAP: South-West Wake county: Elevation NED 10m (elevation@PERMANENT in PERMAN|
|-----------------------------------------------------------------------------|
|                    Category Information                      |   cell|   %  |
|                    #|description                             |  count| cover|
|-----------------------------------------------------------------------------|
|  55.578793-89.162483|from  to . . . . . . . . . . . . . . . .| 329003| 16.25|
| 89.162483-122.746174|from  to . . . . . . . . . . . . . . . .|1075075| 53.09|
|122.746174-156.329865|from  to . . . . . . . . . . . . . . . .| 620922| 30.66|
|-----------------------------------------------------------------------------|
|TOTAL                                                         |2025000|100.00|
+-----------------------------------------------------------------------------+

```

And here is the suggested json output.

```json
 {
    "category": "55.578793-89.162483"
    "description": "from to",
     "range": {
          "from": 55.578793,
          "to": 89.162483
      }
}
```

If the user uses the `-c` flag the category will be `-nan--nan` and the description will be `no data`.

Reply:
fixed.